packages/flutter/lib/src/material/pickers/calendar_date_picker.dart

```diff
@@ -0,0 +1,1050 @@
+    return !_currentDisplayedMonthDate.isBefore(
+      DateTime(widget.lastDate.year, widget.lastDate.month),
+    );
+  }
+
+  void _handleMonthPageChanged(int monthPage) {
+    _setCurrentMonth(utils.addMonthsToMonthDate(widget.firstDate, monthPage));
+  }
+
+  void _handleYearChanged(DateTime value) {
+    _vibrate();
+
+    if (value.isBefore(widget.firstDate)) {
+      value = widget.firstDate;
+    } else if (value.isAfter(widget.lastDate)) {
+      value = widget.lastDate;
+    }
+
+    setState(() {
+      _mode = DatePickerMode.day;
+      _setCurrentMonth(DateTime(value.year, value.month));
+    });
+  }
+
+  void _handleDayChanged(DateTime value) {
+    _vibrate();
+    setState(() {
+      _selectedDate = value;
+      widget.onDateChanged?.call(_selectedDate);
+    });
+  }
+
+  Widget _buildSubheader() {
+    final String previousTooltipText = '${_localizations.previousMonthTooltip} ${_localizations.formatMonthYear(_previousMonthDate)}';
+    final String nextTooltipText = '${_localizations.nextMonthTooltip} ${_localizations.formatMonthYear(_nextMonthDate)}';
+    return _DatePickerSubheader(
+      mode: _mode,
+      nextTooltipText: _isDisplayingLastMonth ? null : nextTooltipText,
+      previousTooltipText: _isDisplayingFirstMonth ? null : previousTooltipText,
+      title: _localizations.formatMonthYear(_currentDisplayedMonthDate),
+      onNextPressed: _isDisplayingLastMonth ? null : _handleNextMonth,
+      onPreviousPressed: _isDisplayingFirstMonth ? null : _handlePreviousMonth,
+      onTitlePressed: () {
+        // Toggle the day/year mode.
+        _handleModeChanged(_mode == DatePickerMode.day ? DatePickerMode.year : DatePickerMode.day);
+      },
+    );
+  }
+
+  Widget _buildPicker() {
+    assert(_mode != null);
+    switch (_mode) {
+      case DatePickerMode.day:
+        return _MonthPicker(
+          key: _pickerKey,
+          currentDate: DateTime.now(),
+          firstDate: widget.firstDate,
+          lastDate: widget.lastDate,
+          selectedDate: _selectedDate,
+          onChanged: _handleDayChanged,
+          onPageChanged: _handleMonthPageChanged,
+          pageController: _pageController,
+          selectableDayPredicate: widget.selectableDayPredicate,
+        );
+      case DatePickerMode.year:
+        return _YearPicker(
+          key: _pickerKey,
+          currentDate: DateTime.now(),
+          firstDate: widget.firstDate,
+          lastDate: widget.lastDate,
+          initialDate: _currentDisplayedMonthDate,
+          selectedDate: _selectedDate,
+          onChanged: _handleYearChanged,
+        );
+    }
+    return null;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final Widget subheader = _buildSubheader();
+    final Widget picker = Flexible(
+      child: SingleChildScrollView(
+        child: SizedBox(
+          height: _maxDayPickerHeight,
+          child: _buildPicker(),
+        ),
+      ),
+    );
+
+    final Widget dialog = OrientationBuilder(
```

Comment:
Why not just return the `OrientationBuilder` here and not bother creating `dialog` at all?

Reply:
At one point I think I was wrapping the dialog with something else that got removed. I will simplify this. Thx.